adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -502,6 +502,11 @@ protected final void onSubscribe(
                     // we do not support subscribing to commands using QoS 2
                     result = Future.failedFuture(new IllegalArgumentException("QoS 2 not supported for command subscription"));
                 } else {
+
+                    if (MqttQoS.AT_LEAST_ONCE.equals(subscription.qualityOfService())) {
```

Comment:
Now we are registering the handler for each topic that the device subscribes to, right? IMHO this means that we won't be able to handle PUBACKs received for messages sent in the context of different topics, right?

I think that we need to register a PUBACK handler at the endpoint level, i.e. a (single) handler that will be responsible for handling *all* PUBACKs we receive from the device. Maybe a new class `Subscriptions` would be helpful in which we keep track of all topics that the device has subscribed to and which can be used as the PUBACK handler ...

Reply:
In order to handle subscription of multiple topics by devices, it makes sense to register PUBACK handler at the endpoint level. Only `msg-id` is available during PUBACKs. I didn't get it how the `Subscriptions` with track of all the subscribed topics would help. IMHO the `msg-id` of the published messages (with Qos 1) along with corresponding subscription info, command context and `timer-id` need to be stored. When any PUBACK is received from the device, the `msg-id`can be used to retrieve the stored values and respond accordingly. WDYT?

Reply:
In this approach with `Subscriptions`,  the topics that are being unsubscribed should also be taken care of. IMHO its more straightforward if we keep track of all the published msg-ids (with Qos1) along with respective CommandSubscription and CommandContext. For example in a class `PublishAcknowledgementHandler` which also can be used as the PUBACK handler. WDYT?
```
class PublishAcknowledgementHandler implements Handler<Integer> {
/**Map<msgId, TriTuple<TimerId, CommandSubscription, CommandContext>>**/
private Map<Integer, TriTuple<Long, CommandSubscription, CommandContext>> publishedWithQoS1AndWaitingForAcknowledgement = new ConcurrentHashMap<>();
```

Reply:
> In this approach with Subscriptions, the topics that are being unsubscribed should also be taken care of.

I agree, that is why I proposed to use it. The Subscriptions class can provide a method that accepts a message ID and handles all PUBACKs of the device. There is no need to implement `Handler<Integer>` because you can simply register the `Subscriptions.handlePubAck` method with the endpoint.

Reply:
@sophokles73 During unsubscribe, the topic entries in Subscriptions class cannot be removed if there are any active acknowledgement timers. Setting a timer (with ack timeout value) and removing the entries should work. In order to avoid such overheads, I have also already worked on a slightly different approach. Can you pls take a quick look at the `AbstractVertxBasedMqttProtocolAdapter` class and provide your comments?

Reply:
I think this goes in the right direction :-)
I wonder if we can add the `CommandConsumer` instance to the `CommandSubscription` class so that we have a handle on the corresponding consumer when a device unsubscribes from a topic.
I also wonder if we can also use `CommandAcknowledgementHandler` to encapsulate the device's subcriptions and then only pass around the `CommandAcknowledgementHandler` instance instead of `CommandAcknowledgementHandler` + `CommandSubscription` ...

Reply:
> I think this goes in the right direction :-)

:+1: 

> I wonder if we can add the `CommandConsumer` instance to the `CommandSubscription` class so that we have a handle on the corresponding consumer when a device unsubscribes from a topic. I would like to get rid of `CommandConnection.closeCommandConsumer()` ...

> I also wonder if we can also use `CommandAcknowledgementHandler` to encapsulate the device's subcriptions and then only pass around the `CommandAcknowledgementHandler` instance instead of `CommandAcknowledgementHandler` + `CommandSubscription` ...

IMHO I would rather keep `CommandAcknowledgementHandler` only to handle the acknowledgement of published commands with Qos 1. Currently it tracks only the published commands (only Qos 1) using MsgId (as map key) and the corresponding  CommandSubscription, CommandContext and TimerIds (as map value).  So it can't be used to keep track of subscriptions or command consumers. For example, a published message (with QoS 0) will have no entry in the  map of `CommandAcknowledgementHandler`.

Reply:
> So it can't be used to keep track of subscriptions or command consumers. For example, a published message (with QoS 0) will have no entry in the map of CommandAcknowledgementHandler.

I know. That's why I think it should be named differently (e.g. `Subscriptions`) and be extended to also contain a list of the device's subscriptions (or just the device's `CommandSubscription` for the time being). It can still also contain the map of message IDs that are pending to be ACKed, but it could also be used to keep track of the subscriptions.

Reply:
@sophokles73 I have named it to `CommandHandler` (since it has handlers to subscribe, unsubscribe and puback for commands). Apart from tracking pubacks, it also tracks all the subscriptions to a device.

> I wonder if we can add the `CommandConsumer` instance to the `CommandSubscription` class so that we have a handle on the corresponding consumer when a device unsubscribes from a topic. I would like to get rid of `CommandConnection.closeCommandConsumer()` ...

This `CommandHandler` also contains reference to the `CommandConsumer` instances, so I am able to get rid of the `CommandConnection.closeCommandConsumer()`